File: src/player_registry.py
"""Player registry for caching confirmed player matches."""

import json
from datetime import datetime
from pathlib import Path
from typing import Dict, Optional
from dataclasses import dataclass, asdict

from .config import debug_log


@dataclass
class RegisteredPlayer:
    """A player stored in the registry."""
    dupr_id: str
    dupr_name: str
    rating: Optional[float]
    location: Optional[str]
    last_updated: str
    
    def to_dict(self) -> dict:
        """Convert to dictionary for JSON serialization."""
        return asdict(self)
    
    @classmethod
    def from_dict(cls, data: dict) -> 'RegisteredPlayer':
        """Create from dictionary."""
        return cls(
            dupr_id=data.get('dupr_id', ''),
            dupr_name=data.get('dupr_name', ''),
            rating=data.get('rating'),
            location=data.get('location'),
            last_updated=data.get('last_updated', '')
        )


class PlayerRegistry:
    """Registry for caching confirmed player matches.
    
    Stores mappings from search names to DUPR player info,
    allowing instant lookups for previously matched players.
    """
    
    def __init__(self, registry_file: Optional[str] = None):
        """Initialize the registry.
        
        Args:
            registry_file: Path to the registry JSON file. If None, uses default location.
        """
        if registry_file is None:
            project_root = Path(__file__).parent.parent
            registry_file = project_root / "config" / "player_registry.json"
        
        self._filepath = Path(registry_file)
        self._registry: Dict[str, RegisteredPlayer] = {}
        self._dirty = False  # Track if changes need to be saved
        
        self._load()
    
    def _normalize_key(self, name: str) -> str:
        """Normalize a name for use as a registry key."""
        return name.lower().strip()
    
    def _load(self) -> None:
        """Load the registry from file."""
        if not self._filepath.exists():
            debug_log(f"Player registry not found at {self._filepath}")
            return
        
        try:
            with open(self._filepath, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            for name, player_data in data.items():
                self._registry[self._normalize_key(name)] = RegisteredPlayer.from_dict(player_data)
            
            debug_log(f"Loaded {len(self._registry)} players from registry")
            
        except (json.JSONDecodeError, IOError) as e:
            debug_log(f"Error loading player registry: {e}")
    
    def save(self) -> None:
        """Save the registry to file."""
        if not self._dirty:
            return
        
        try:
            # Ensure directory exists
            self._filepath.parent.mkdir(parents=True, exist_ok=True)
            
            # Convert to serializable format
            data = {name: player.to_dict() for name, player in self._registry.items()}
            
            with open(self._filepath, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2)
            
            self._dirty = False
            debug_log(f"Saved {len(self._registry)} players to registry")
            
        except IOError as e:
            debug_log(f"Error saving player registry: {e}")
    
    def get(self, search_name: str) -> Optional[RegisteredPlayer]:
        """Get a player from the registry by search name.
        
        Args:
            search_name: The name that was used to search for the player.
            
        Returns:
            RegisteredPlayer if found, None otherwise.
        """
        key = self._normalize_key(search_name)
        return self._registry.get(key)
    
    def register(
        self,
        search_name: str,
        dupr_id: str,
        dupr_name: str,
        rating: Optional[float] = None,
        location: Optional[str] = None
    ) -> None:
        """Register a player mapping.
        
        Args:
            search_name: The name used to search for the player.
            dupr_id: The DUPR ID of the matched player.
            dupr_name: The full name in DUPR.
            rating: The player's rating.
            location: The player's location.
        """
        key = self._normalize_key(search_name)
        
        player = RegisteredPlayer(
            dupr_id=dupr_id,
            dupr_name=dupr_name,
            rating=rating,
            location=location,
            last_updated=datetime.now().isoformat()
        )
        
        self._registry[key] = player
        self._dirty = True
        
        debug_log(f"Registered '{search_name}' -> '{dupr_name}' ({dupr_id})")
    
    def contains(self, search_name: str) -> bool:
        """Check if a name is in the registry.
        
        Args:
            search_name: The name to check.
            
        Returns:
            True if the name is registered.
        """
        return self._normalize_key(search_name) in self._registry
    
    def remove(self, search_name: str) -> bool:
        """Remove a player from the registry.
        
        Args:
            search_name: The name to remove.
            
        Returns:
            True if the player was removed, False if not found.
        """
        key = self._normalize_key(search_name)
        if key in self._registry:
            del self._registry[key]
            self._dirty = True
            return True
        return False
    
    def __len__(self) -> int:
        """Return the number of registered players."""
        return len(self._registry)
    
    def __contains__(self, search_name: str) -> bool:
        """Check if a name is in the registry."""
        return self.contains(search_name)


# Global instance for convenience
_registry: Optional[PlayerRegistry] = None


def get_registry() -> PlayerRegistry:
    """Get the global PlayerRegistry instance."""
    global _registry
    if _registry is None:
        _registry = PlayerRegistry()
    return _registry


def save_registry() -> None:
    """Save the global registry to file."""
    if _registry is not None:
        _registry.save()


def reset_registry() -> None:
    """Reset the global registry (useful for testing)."""
    global _registry
    _registry = None

File: packages/core/src/player-registry.ts
/**
 * Player Registry Module
 * Caches player lookups to avoid repeated API calls for known players
 */

export interface RegisteredPlayer {
  /** Original search name used to find this player */
  searchName: string;
  /** DUPR system ID */
  duprId: string;
  /** Full name as shown in DUPR */
  duprName: string;
  /** Player's best rating (doubles preferred) */
  rating: number | null;
  /** Player's location from DUPR */
  location: string;
  /** When this player was registered */
  registeredAt: Date;
}

export interface RegistryData {
  [searchName: string]: {
    duprId: string;
    duprName: string;
    rating: number | null;
    location: string;
    registeredAt: string;
  };
}

export class PlayerRegistry {
  private registry: Map<string, RegisteredPlayer>;
  private filePath: string | null;

  /**
   * Create a new PlayerRegistry
   * @param filePath Optional path to JSON file for persistence
   */
  constructor(filePath?: string) {
    this.registry = new Map();
    this.filePath = filePath ?? null;
  }

  /**
   * Normalize a search name for consistent lookups
   * - Convert to lowercase
   * - Trim whitespace
   */
  private normalizeKey(name: string): string {
    return name.toLowerCase().trim();
  }

  /**
   * Register a player in the cache
   * @param searchName The name used to search for this player
   * @param duprId DUPR system ID
   * @param duprName Full name from DUPR
   * @param rating Player's rating (null if not rated)
   * @param location Player's location
   */
  register(
    searchName: string,
    duprId: string,
    duprName: string,
    rating: number | null,
    location: string
  ): void {
    const key = this.normalizeKey(searchName);
    this.registry.set(key, {
      searchName,
      duprId,
      duprName,
      rating,
      location,
      registeredAt: new Date(),
    });
  }

  /**
   * Get a registered player by search name
   * @param searchName The name to look up
   * @returns RegisteredPlayer if found, undefined otherwise
   */
  get(searchName: string): RegisteredPlayer | undefined {
    const key = this.normalizeKey(searchName);
    return this.registry.get(key);
  }

  /**
   * Check if a player is registered
   * @param searchName The name to check
   * @returns true if player is registered
   */
  has(searchName: string): boolean {
    const key = this.normalizeKey(searchName);
    return this.registry.has(key);
  }

  /**
   * Remove a player from the registry
   * @param searchName The name to remove
   * @returns true if player was removed
   */
  delete(searchName: string): boolean {
    const key = this.normalizeKey(searchName);
    return this.registry.delete(key);
  }

  /**
   * Get all registered players
   * @returns Array of all registered players
   */
  getAll(): RegisteredPlayer[] {
    return Array.from(this.registry.values());
  }

  /**
   * Get the number of registered players
   */
  get size(): number {
    return this.registry.size;
  }

  /**
   * Clear all registered players
   */
  clear(): void {
    this.registry.clear();
  }

  /**
   * Convert registry to JSON-serializable format
   */
  toJSON(): RegistryData {
    const data: RegistryData = {};
    for (const [key, player] of this.registry) {
      data[key] = {
        duprId: player.duprId,
        duprName: player.duprName,
        rating: player.rating,
        location: player.location,
        registeredAt: player.registeredAt.toISOString(),
      };
    }
    return data;
  }

  /**
   * Load registry from JSON data
   * @param data Registry data object
   */
  fromJSON(data: RegistryData): void {
    this.registry.clear();
    for (const [key, value] of Object.entries(data)) {
      this.registry.set(key, {
        searchName: key,
        duprId: value.duprId,
        duprName: value.duprName,
        rating: value.rating,
        location: value.location,
        registeredAt: new Date(value.registeredAt),
      });
    }
  }

  /**
   * Save registry to file (Node.js environment)
   * Note: This is a no-op in browser environments
   */
  async save(): Promise<void> {
    if (!this.filePath) return;

    // Dynamic import for Node.js fs module
    try {
      const fs = await import('fs/promises');
      const data = JSON.stringify(this.toJSON(), null, 2);
      await fs.writeFile(this.filePath, data, 'utf-8');
    } catch (error) {
      // In browser or if fs not available, silently skip
      console.warn('Unable to save registry to file:', error);
    }
  }

  /**
   * Load registry from file (Node.js environment)
   * Note: This is a no-op in browser environments
   */
  async load(): Promise<void> {
    if (!this.filePath) return;

    try {
      const fs = await import('fs/promises');
      const data = await fs.readFile(this.filePath, 'utf-8');
      const json = JSON.parse(data) as RegistryData;
      this.fromJSON(json);
    } catch (error) {
      // File doesn't exist or can't be read - start with empty registry
      console.warn('Unable to load registry from file:', error);
    }
  }
}
